### PR TITLE
Add layout + compute_layout support: TransformerNMT, BERT, ALBERT, ELECTRA, MobileBERT, RoBERTA, XLMR

### PR DESCRIPTION
**Pull request by sxjscience**
@MoisesHer @dmlc/gluon-nlp-team 

This PR adds two additional flags to backbone models to enhance the computational speed and usability.

- layout
The layout of the inputs + outputs, where
   - NT: (batch_size, sequence_length, ...)
   - TN: (sequence_length, batch_size, ...)
- compute_layout
The layout of the inner computation. By default, we have the `"auto"` option, in which GluonNLP will determine the best layout based on heuristics.

The technical insights about why layouts may matter lies as follows (also documented here https://github.com/dmlc/gluon-nlp/blob/cd48efdd51cf26b47791c0329e1092fb5fc658f7/src/gluonnlp/attention_cell.py#L540-L546):

When the layouts of memory and query are "TNC", they have the shape:
- query: `(L_query, B, N, C_Q)`
- memory: `(L_mem, B, N, C_Q)`

One step in the AttentionCell is to obtain the multi-head attention scores, which can be written as follows:

`(L_query, B, N, C_Q) X (L_mem, B, N, C_Q) -> (B, N, L_query, L_mem)`

This layout structure can be implemented very efficiently because B, N are consecutive to each other. 

To have a clear picture of what's happening, we may consider the (i, j)th element of the output, i.e.,
```
out[i, j, :, :] = query[:, i, j, :] X key[:, i, j, :].T
```
This is just one GEMM call. We can thus implement the **whole kernel** via a single call of batched GEMM by correctly specifying the strides.

Also, in fairseq, the inner computation of the TransformerEncoder is using the TN layout:

https://github.com/pytorch/fairseq/blob/108bb2560b1ec01524ba723bc7c69186875afa0a/fairseq/models/transformer.py#L399-L407

After this PR, these models will have the layout flag:
- [x] TransformerEncoder, TransformerDecoder, TransformerNMTModel
- [x] BertModel
- [x] AlbertModel
- [x] ElectraModel
- [x] MobileBertModel
- [x] RobertaModel
- [x] XLMRModel
- [x] Test cases

### CODE REVIEW

**Comment by sxjscience** [comment]
Now I'm waiting for https://github.com/dmlc/gluon-nlp/pull/1261. We added two flags:
- layout
It means the layout of the input + output
- compute_layout
It means the layout of the inner computation. By default, we will always use `compute_layout=auto`, which will automatically determine the best compute layout based on our heuristic.

File: src/gluonnlp/models/albert.py [hunk 1]
```diff
@@ -38,37 +39,109 @@
...
+    # Hyper-parameters of the Initializers
+    cfg.INITIALIZER = CN()
+    cfg.INITIALIZER.embed = ['truncnorm', 0, 0.02]
+    cfg.INITIALIZER.weight = ['truncnorm', 0, 0.02]  # TruncNorm(0, 0.02)
+    cfg.INITIALIZER.bias = ['zeros']
+    # Version of the model. This helps ensure backward compatibility.
+    # Also, we can not use string here due to https://github.com/rbgirshick/yacs/issues/26
+    cfg.VERSION = 1
+    cfg.freeze()
+    return cfg
```
Comment (sxjscience) [inline src/gluonnlp/models/albert.py:82]: @ZheyuYe I'm moving the cfgs to the file to make is consistent to Transformer + RoBERTA.

File: src/gluonnlp/models/albert.py [hunk 2]
```diff
@@ -314,24 +422,37 @@ def get_initial_embedding(self, F, inputs, token_types=None):
...
         -------
         embedding
             The initial embedding that will be fed into the encoder
+            - layout = 'NT'
+                Shape (batch_size, seq_length, C_embed)
+            - layout = 'TN'
+                Shape (seq_length, batch_size, C_embed)
         """
+        if self.layout == 'NT':
+            time_axis, batch_axis = 1, 0
```
Comment (ZheyuYe) [inline src/gluonnlp/models/albert.py:445]: would this be more intuitive by swapping the order like `batch_axis, time_axis = 0, 1`

File: src/gluonnlp/attention_cell.py [hunk 3]
```diff
@@ -139,15 +152,17 @@ def gen_self_attn_mask(F, data,
             batch_mask = (F.np.expand_dims(steps, axis=0) < F.np.expand_dims(valid_length, axis=-1)).astype(dtype)
             mask = mask * F.np.expand_dims(batch_mask, axis=-1)
         else:
-            batch_ones = F.np.ones_like(F.npx.arange_like(data, axis=0), dtype=np.float32)  # (batch_size,)
+            batch_ones = F.np.ones_like(F.npx.arange_like(data, axis=batch_axis),
+                                        dtype=np.float32)  # (batch_size,)
```
Comment (ZheyuYe) [inline src/gluonnlp/attention_cell.py:156]: dtype= dtype

File: src/gluonnlp/models/albert.py (hunk 1 above)
Comment (ZheyuYe) [reply src/gluonnlp/models/albert.py:82]: ok, I am going to revise other models

File: tests/test_models_bert.py [hunk 4]
```diff
@@ -12,6 +11,43 @@ def test_list_pretrained_bert():
...
+    token_types = mx.np.random.randint(0, 2, (batch_size, sequence_length))
+    valid_length = mx.np.random.randint(3, sequence_length, (batch_size,))
+    masked_positions = mx.np.random.randint(0, 3, (batch_size, num_mask))
+
+    # Test for BertModel
+    bert_model = BertModel.from_cfg(cfg)
+    bert_model.initialize()
+    bert_model.hybridize()
+    contextual_embedding, pooled_out = bert_model(inputs, token_types, valid_length)
+    bert_model_tn = BertModel.from_cfg(cfg_tn, params=bert_model.collect_params())
```
Comment (ZheyuYe) [inline tests/test_models_bert.py:42]: `share_parameters`

File: src/gluonnlp/models/albert.py (hunk 1 above)
Comment (ZheyuYe) [reply src/gluonnlp/models/albert.py:82]: ok, I am going to revise other models

File: README.md [hunk 5]
```diff
@@ -21,10 +21,10 @@ First of all, install the latest MXNet. You may use the following commands:
 ```bash
 
 # Install the version with CUDA 10.1
-pip install -U --pre mxnet-cu101>=2.0.0b20200604 -f https://dist.mxnet.io/python
+pip install -U --pre "mxnet-cu101>=2.0.0b20200604,<=2.0.0b20200619" -f https://dist.mxnet.io/python
```
[the remark this answers is not in the document]
Comment (szha) [reply README.md:24]: cu101 support is discontinued and we will need to switch to cu102 for later commits. cc @leezu to confirm

File: tests/test_models_bert.py (hunk 4 above)
Comment (ZheyuYe) [reply tests/test_models_bert.py:42]: This is the reason why CI did not pass, could you plz fix it?

File: src/gluonnlp/models/electra.py [hunk 6]
```diff
@@ -96,6 +163,7 @@ def get_generator_cfg(model_config):
 FILE_STATS = load_checksum_stats(os.path.join(get_model_zoo_checksum_dir(), 'electra.txt'))
 
 
+# TODO(sxjscience) Use BertTransformer
```
Comment (sxjscience) [inline src/gluonnlp/models/electra.py:166]: @ZheyuYe I think we may try to directly reuse BertTransformer. Do you think it's possible?

Comment (ZheyuYe) [reply src/gluonnlp/models/electra.py:166]: Yeah we could do that, encoders in bert and electra are almost same.

Comment (sxjscience) [reply src/gluonnlp/models/electra.py:166]: OK, so let me keep the TODO and we can revise later.

**Comment by sxjscience** [comment]
@dmlc/gluon-nlp-committers  @MoisesHer @ZheyuYe   Should be ready for review.

File: src/gluonnlp/attention_cell.py (hunk 3 above)
Comment (sxjscience) [reply src/gluonnlp/attention_cell.py]: solved

File: src/gluonnlp/models/albert.py (hunk 2 above)
Comment (sxjscience) [reply src/gluonnlp/models/albert.py:447]: solved

File: tests/test_models_bert.py (hunk 4 above)
Comment (sxjscience) [reply tests/test_models_bert.py]: fixed

**Pull request merged**